### PR TITLE
8329817: Augment prototype Float16 class

### PR DESCRIPTION
**Pull request by jddarcy**
Sketch of a more full-featured API for Float16.

---------
### Progress
- [x] Change must not contain extraneous whitespace

### Issue
 * [JDK-8329817](https://bugs.openjdk.org/browse/JDK-8329817): Augment prototype Float16 class (**Enhancement** - P4)

### Reviewers
 * [Jatin Bhateja](https://openjdk.org/census#jbhateja) (@jatin-bhateja - Committer) ⚠️ Review applies to [bf09c544](https://git.openjdk.org/valhalla/pull/1117/files/bf09c544a2b365f1e5c813d5eed22471b57d1d25)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/valhalla.git pull/1117/head:pull/1117` \
`$ git checkout pull/1117`

Update a local copy of the PR: \
`$ git checkout pull/1117` \
`$ git pull https://git.openjdk.org/valhalla.git pull/1117/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1117`

View PR using the GUI difftool: \
`$ git pr show -t 1117`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/valhalla/pull/1117.diff">https://git.openjdk.org/valhalla/pull/1117.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/valhalla/pull/1117#issuecomment-2153916698)

### CODE REVIEW

**Comment by jddarcy** [comment]
This PR has some known to-do items, including testing, but does provide a more full featured API set on par with Float or Double.

One to-do item is writing the double -> float16 conversion routine. Other items include adjusting the binary <-> decimal conversion routines to deal with the reduced precision of float16.

File: src/java.base/share/classes/java/lang/Float16.java [hunk 1]
```diff
@@ -54,8 +57,123 @@
 
 // Currently Float16 is a primitive class but in future will be aligned with
 // Enhanced Primitive Boxes described by JEP-402 (https://openjdk.org/jeps/402)
-public primitive class Float16 extends Number {
+public /*primitive*/ class Float16 extends Number
```
[the remark this answers is not in the document]
Comment (jddarcy) [reply src/java.base/share/classes/java/lang/Float16.java:60]: I don't know the proper idiom for that use. (I commented out the "primitive" modifier so I could get the doc build to work for the class.)

Do you know what the plans are for syncing in changes from Valhalla mainline into lworld+fp16?

[the remark this answers is not in the document]
Comment (jatin-bhateja) [reply src/java.base/share/classes/java/lang/Float16.java:60]: Hi @jddarcy , I have just integrated a [merge pull request](https://github.com/openjdk/valhalla/pull/1120) to merge latest code from lworld. With this I made Floar16 an implicitly constructible value based class in order to create flat layout representation for Float16 arrays.

[the remark this answers is not in the document]
Comment (jddarcy) [reply src/java.base/share/classes/java/lang/Float16.java]: Thanks @jatin-bhateja . I merged in the changes and updated the Float16 tests to pass.

**Comment by jatin-bhateja** [comment]
Hi @jddarcy , 

There are few more tests which may need refinement 
jtreg:test/hotspot/jtreg/compiler/intrinsics/float16/TestFP16ScalarAdd.java
jtreg:test/hotspot/jtreg/compiler/vectorization/TestFloat16VectorSum.java

Best Regards,
Jatin

**Comment by jatin-bhateja** [comment]
Hey @jddarcy,

Taking the liberty to append some notes on this PR for future extensions for flat layout representation. Your comments and suggestion will be highly appreciated.

Q. What will it take to support a flat layout for an array of Float16 values ?

A.  I will begin by mentioning some basic terminology related to nullity restrictions.

1) ImpliciltyConstructible :  An instance which is always assigned a default value and guarantees non-nullability by construction.
2) NullRestricted              :  A value field if null restricted will never be assigned a null value during its lifetime.  

Another question which I wanted to answer here is about representing a null value, a value object if not null restricted must be able to encode a null value in some form, for example, a value class with just one 32 bit integer field can only accommodate integral value range in 2s complement representation, in order to encode a null value it may need additional bit, thus a 32 bit integer with nullity may need more than 32 bits storage. Thus with a mix of null and non-null values flattening may not be much useful due to heterogeneity in element sizes. To simplify layout flattening we can always assume uniform element size by  always assuming null encoding, but then it may add complexities in auto-vectorizing these contiguously placed values.  

With above definitions in place,  it becomes easy for runtime to generate a flat memory layout for arrays of Float16 types if implicit constructability and non-nullability are always guaranteed.

Existing ValueBased classes under --enable-preview option ensures generating InlineKlass. 

Since we are making Float16 a value based class, hence we can extend the scope of value based classes to imply 1) and 2), but its a behavioral change and may impact semantics of existing value classes. For the time being we can decorate Float16 with 1) and 2) annotations along with some minor modifications in executions engines. to ensure flat layout.

Only doubt I have is w.r.t to application of NullRestricted annotation to an array, since its target type is a FIELD, thus it technically does not cover an array element.  

I have created a follow-up JBS for facilitating flat array layout for Float16 values.
https://bugs.openjdk.org/browse/JDK-8333852

Best Regards,
Jatin

**Comment by jddarcy** [comment]
> Hi @jddarcy ,
> 
> There are few more tests which may need refinement jtreg:test/hotspot/jtreg/compiler/intrinsics/float16/TestFP16ScalarAdd.java jtreg:test/hotspot/jtreg/compiler/vectorization/TestFloat16VectorSum.java
> 
> Best Regards, Jatin

Okay; I updated both tests to reflect the changes in the Float16 API. While TestFloat16VectorSum.java run successfully, TestFP16ScalarAdd fails at runtime with a message equivalent to "preview not enabled." Given the source of the test, I'm not sure what preview isn't being enabled as needed.

**Comment by jddarcy** [comment]
A few comments on the proposed revised API of Float16:

The Float16 class was generally brought up to static field and method level parity with java.lang.{Float, Double}. Additionally, a number of low-level floating-point handling methods from Math/StrictMath were added to, such as getExponent. These methods are usually an operation or recommended function from IEEE 754. Methods for arithmetic operators {+, -, *, /} and square root were also added. The value(short) method was changed from a bit-level conversion to a value-based one; bit-level conversion is provided by a new shortBitsToFloat16 method.

The current API favors static methods over instances one. Methods to extend the Number abstract class and implement Comparable are of necessity instance methods.

Some more work is needed on the decimal <-> binary conversion methods.

**Comment by jddarcy** [comment]
> A few comments on the proposed revised API of Float16:
> 
> The Float16 class was generally brought up to static field and method level parity with java.lang.{Float, Double}. Additionally, a number of low-level floating-point handling methods from Math/StrictMath were added to, such as getExponent. These methods are usually an operation or recommended function from IEEE 754. Methods for arithmetic operators {+, -, *, /} and square root were also added. The value(short) method was changed from a bit-level conversion to a value-based one; bit-level conversion is provided by a new shortBitsToFloat16 method.
> 
> The current API favors static methods over instances one. Methods to extend the Number abstract class and implement Comparable are of necessity instance methods.
> 
> Some more work is needed on the decimal <-> binary conversion methods.

PS As a design note on the implementation, as many of the methods are expected to be intrinsified on common production platforms, I was looking to favor conciseness and code reuse, say by relying on float and double operations, rather than absolute performance for a native software-based implementation of Float16.

**Review by jatin-bhateja — APPROVED** [review]
Hi @jddarcy , Apart from few minor comments, patch looks good to me, there is a build error due to malformed javadoc comment. 

Kindly fix and integrate.

**Comment by jddarcy** [comment]
> Hi @jddarcy , Apart from few minor comments, patch looks good to me, there is a build error due to malformed javadoc comment.
> 
> Kindly fix and integrate.

Oops; sorry for the build breakage -- no change is no small that you shouldn't build before pushing ;-)

File: src/java.base/share/classes/java/lang/Float16.java [hunk 2]
```diff
@@ -48,6 +51,16 @@
...
+ * @apiNote
+ * The methods in this class generally have analogous methods in
+ * either {@link Float}/{@link Double} or {@link Math}/{@link
+ * StrictMath}. Unless otherwise specified, the handling of special
+ * floating-point values such as {@linkplain #isNaN(Float16) NaN}
+ * values, {@linkplain #isInfinite(Float16) infinities}, and signed
+ * zeros of methods in this class is wholly analogous to the handling
+ * of equivalent cases by methods in {@code Float}, {@code Double},
+ * {@code Math}, etc.
+ *
```
[the remark this answers is not in the document]
Comment (jddarcy) [reply src/java.base/share/classes/java/lang/Float16.java:63]: In the mainline JDK, we've long since stopped adding new `@author` tags. (I've actually removed some of my `@author` tags in some test code.) In preparation for a future where Float16 is in the mainline JDK, I suggest this information be recorded in some other fashion, at least when integrated into the mainline.

The `/contributor ` Skara command can be used to record multiple co-authors of the code in a PR. That is one way this information could be preserved when integrated into the mainline.

**Comment by jddarcy** [comment]
> Hi @jddarcy , Apart from few minor comments, patch looks good to me, there is a build error due to malformed javadoc comment.
> 
> Kindly fix and integrate.

Thanks; let me take a pass at writing at least some basic regression tests before pushing.

@jatin-bhateja , do you know if promoting the three operands of a Float16 fma to double, doing the operation in double, and rounding to Float16 is sufficient to correctly implement a Float16 fma? I haven't worked through all the cases yet and I'm not certain they're cannot be double-rounding issues. (If double rounding turns out to be a problem, I was thinking it would be possible to see if (a*b + c) was exact in double, and if, not add in a sticky bit to make sure the rounding occurs properly, but I haven't developed the details yet.)

**Comment by jatin-bhateja** [comment]
> > Hi @jddarcy , Apart from few minor comments, patch looks good to me, there is a build error due to malformed javadoc comment.
> > Kindly fix and integrate.
> 
> Thanks; let me take a pass at writing at least some basic regression tests before pushing.
> 
> @jatin-bhateja , do you know if promoting the three operands of a Float16 fma to double, doing the operation in double, and rounding to Float16 is sufficient to correctly implement a Float16 fma? I haven't worked through all the cases yet and I'm not certain they're cannot be double-rounding issues. (If double rounding turns out to be a problem, I was thinking it would be possible to see if (a*b + c) was exact in double, and if, not add in a sticky bit to make sure the rounding occurs properly, but I haven't developed the details yet.)

Hi @jddarcy , As per specification of [Math.fma(float, float, float)](https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/lang/Math.java#L2494) internal computation of constituent operation (mul and add) should be done at infinite precision and only final result should be rounded, we are now upcasting float16 to double but it will not prevent rounding happening for mul and add.

File: src/java.base/share/classes/java/lang/Float16.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jddarcy) [reply src/java.base/share/classes/java/lang/Float16.java:67]: Hi @jatin-bhateja,
With the latest commit, I think the spec is close to the spec that would be acceptable to include in mainline. Please confirm you're okay with this state of the PR being integrated. Afterward, work can continue refining a few areas binary <-> decimal spec and implementation, fma internal comment and/or implementation, etc. Thanks.

**Comment by jddarcy** [comment]
> > > Hi @jddarcy , Apart from few minor comments, patch looks good to me, there is a build error due to malformed javadoc comment.
> > > Kindly fix and integrate.
> > 
> > 
> > Thanks; let me take a pass at writing at least some basic regression tests before pushing.
> > @jatin-bhateja , do you know if promoting the three operands of a Float16 fma to double, doing the operation in double, and rounding to Float16 is sufficient to correctly implement a Float16 fma? I haven't worked through all the cases yet and I'm not certain they're cannot be double-rounding issues. (If double rounding turns out to be a problem, I was thinking it would be possible to see if (a*b + c) was exact in double, and if, not add in a sticky bit to make sure the rounding occurs properly, but I haven't developed the details yet.)
> 
> Hi @jddarcy , As per specification of [Math.fma(float, float, float)](https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/lang/Math.java#L2494) internal computation of constituent operation (mul and add) should be done at infinite precision and only final result should be rounded, we are now upcasting float16 to double but it will not prevent rounding happening for mul and add.

Right; the fma needs to operate as-if it used infinite precision internally. This could be implemented (slowly) using JDK classes by implementing a BigDecimal -> Float16 conversion after doing a*b+c in BigDeicmal. I've been considering adding BigDecimal -> Float16 conversion anyway for completeness in the platform.

My understanding of how fma is implemented in hardware is that for a format with P bits of precision, there is a ~2P wide internal register to hold the exact product as an intermediate result. Then the value being adding in can be aligned at the right exponent location and the final rounding back to P bits of precision can occur, with logic of any sticky bit for rounding to nearest even, etc.

There are many cases where double (P = 53) will exactly hold the product and sum of three Float16 (P = 11) operands.  However, the product can be so large or so small that rounding  occurs when the third operand is added in.

I haven't worked through if the potential round-offs are all benign with the final rounding to Float16 or if some corrective action would need to be taken to get the effect of a sticky bit. For example, if a*b is so large that the highest exponent position set is more than 53 positions away from the lowest exponent position set on c, if the final result is going to overflow anyway, the round-off in computing a*b+c in double doesn't matter. It might be problematic if a*b is much smaller than c, but that is another case I haven't fully thought through yet.

I'll give an updated on my analysis/research on this fma issue by next week.

File: src/java.base/share/classes/java/lang/Float16.java [hunk 3]
```diff
@@ -56,129 +73,941 @@
...
+     *
+     * @param addend the first operand
+     * @param augend the second operand
+     * @return the sum of the operands
+     *
+     * @jls 15.4 Floating-point Expressions
      */
-    public short float16ToRawShortBits() { return value; }
+    // @IntrinsicCandidate
+    public static Float16 add(Float16 addend, Float16 augend) {
```
Comment (Bhavana-Kilambi) [inline src/java.base/share/classes/java/lang/Float16.java:736]: Hi, thanks for implementing these operations for Float16. May I ask why do we have both `Float16 sum()` and `Float16 add()`? Also, both seem to be marked as `@IntrinsicsCandidate`.

Comment (jddarcy) [reply src/java.base/share/classes/java/lang/Float16.java:736]: > Hi, thanks for implementing these operations for Float16. May I ask why do we have both `Float16 sum()` and `Float16 add()`? Also, both seem to be marked as `@IntrinsicsCandidate`.

Hello @Bhavana-Kilambi ,
The Float16 class already had a sum method before I started my work, so I left it in place. I assume "sum" was added in analogy to the the sum method added to Float and Double in the context of the stream work done in JDK 8.

For the purpose of including support for methods to perform the addition, subtraction, multiplication, and division operation, I think "add" is a better name than "sum".

I don't have any appreciable experience changing HotSpot, so I was going to leave that aspect of this work to others. The build fails if the `@IntrinsicCandidate` annotation is applied to methods without the possibility of a matching candidate in HotSpot. Therefore, I left a commented out form of the annotation on methods I assumed would be eligible for such support, but those were just my guesses and should be updated for what the actual intrinsic support could be.

Comment (Bhavana-Kilambi) [reply src/java.base/share/classes/java/lang/Float16.java:736]: Ok, got it. Thank you for the detailed response.

File: test/hotspot/jtreg/compiler/vectorization/TestFloat16VectorSum.java [hunk 4]
```diff
@@ -33,7 +33,7 @@
 package compiler.vectorization;
```
Comment (Bhavana-Kilambi) [inline test/hotspot/jtreg/compiler/vectorization/TestFloat16VectorSum.java:33]: I am not able to add comments to the lines above this line. I see "-XDenablePrimitiveClasses" flag on line #29 which I think is obsolete after the latest merge. If this test passes for you then you can leave this unchanged. I can change this line in my next PR. Thanks.

**Comment by jddarcy** [comment]
> > > > Hi @jddarcy , Apart from few minor comments, patch looks good to me, there is a build error due to malformed javadoc comment.
> > > > Kindly fix and integrate.
> > > 
> > > 
> > > Thanks; let me take a pass at writing at least some basic regression tests before pushing.
> > > @jatin-bhateja , do you know if promoting the three operands of a Float16 fma to double, doing the operation in double, and rounding to Float16 is sufficient to correctly implement a Float16 fma? I haven't worked through all the cases yet and I'm not certain they're cannot be double-rounding issues. (If double rounding turns out to be a problem, I was thinking it would be possible to see if (a*b + c) was exact in double, and if, not add in a sticky bit to make sure the rounding occurs properly, but I haven't developed the details yet.)
> > 
> > 
> > Hi @jddarcy , As per specification of [Math.fma(float, float, float)](https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/lang/Math.java#L2494) internal computation of constituent operation (mul and add) should be done at infinite precision and only final result should be rounded, we are now upcasting float16 to double but it will not prevent rounding happening for mul and add.
> 
> Right; the fma needs to operate as-if it used infinite precision internally. This could be implemented (slowly) using JDK classes by implementing a BigDecimal -> Float16 conversion after doing a*b+c in BigDeicmal. I've been considering adding BigDecimal -> Float16 conversion anyway for completeness in the platform.
> 
> My understanding of how fma is implemented in hardware is that for a format with P bits of precision, there is a ~2P wide internal register to hold the exact product as an intermediate result. Then the value being adding in can be aligned at the right exponent location and the final rounding back to P bits of precision can occur, with logic of any sticky bit for rounding to nearest even, etc.
> 
> There are many cases where double (P = 53) will exactly hold the product and sum of three Float16 (P = 11) operands. However, the product can be so large or so small that rounding occurs when the third operand is added in.
> 
> I haven't worked through if the potential round-offs are all benign with the final rounding to Float16 or if some corrective action would need to be taken to get the effect of a sticky bit. For example, if a_b is so large that the highest exponent position set is more than 53 positions away from the lowest exponent position set on c, if the final result is going to overflow anyway, the round-off in computing a_b+c in double doesn't matter. It might be problematic if a*b is much smaller than c, but that is another case I haven't fully thought through yet.
> 
> I'll give an updated on my analysis/research on this fma issue by next week.

An update: promoting the three float16 fma operands to double and then computing (a*b) + c in double and rounding once at the end to Float16 works for (at least) most possible operands to a Float16 fma. It may work for all operands; I'm still working through those details.

To summarize my findings. if the product-sum held in double is exact, the single rounding to Float16 will compute the correct result.

Each Float16 operand is exactly convertible to double and the a*b product is exact since double has more than twice the precision of Float16. That leaves analyzing whether or not the product-sum is exact.

For Float16 values, the possible exponent bit positions that could be set range from 2^15 (from MAX_VALUE) to 2^(-24) (from MIN_VALUE). This entire exponent range of 15 - (-24) + 1 = 40 is less than the _precision_ of double, 53 bits.

When multiplying two Float16 values together, the exponent bit positions set in the exact product range over 2^31 (MAX_VALUE squared) to 2^(-48) (MIN_VALUE) squared.

If the exact product is larger than about 2^16, the final rounded result must overflow Float16. If the exact product has a highest exponent bit of 2^16, the product-sum will be exact since a double could hold a value as small as 2^(16 - 52) = 2^(-36), which is smaller than Float16.MIN_VALUE. Therefore, when the product has an exponent of 2^16 or larger, the end result is correct after conversion to Float16.

A similar argument holds when the product has an exponent in the normal range of Float16; any additional sum will be exact given the precision of double.

However, if the product is sufficiently small, it is possible the double will not be able to hold the exact value since the exponent range of 2^(-48) to 2^(15) is larger than double precision. However, for many operands, even if the product has underflowed, adding in many operands may be exact in practice. This exactness can be tested for using the 2sum algorithm (https://en.wikipedia.org/wiki/2Sum) for a zero trailing "t" component.

Fortunately in terms of computing the final answer and rounding complications, if c is the subnormal range of Float16 (exponent range 2^(-14) to 2(-24)), the product-sum will be exactly representable in double.

That leaves possibly non-exact product-sum with a combination of product in the subnormal range of Float16 and the c term to be added in being not-small. However, if this product-sum is non-exact, the smaller term from the product, with at most 22 exponent bit positions set, and the the 11 bits from c being summed in, must be separated by at least 53 - (22 + 11) = 20 bit positions otherwise the product-sum would fit in a double. I believe this implies at least one of the double-rounding scenarios cannot occur, in particular a half-way result in the smaller precision, Float16 in this case, rounding differently because sticky bit information from the higher precision was rounded away.

I'll keep working through the non-exact cases. Assuming a `Float16 valueOf(BigDecimal)` method is added, an "obviously right" approach would be to use a*b + c in double for all exact cases and failover to computing the exact result in BigDecimal and then rounding to that to Float16 for all the inexact cases.

**Comment by jddarcy** [comment]
> Why separate entry points for `add` and `sum`?

See previous comment https://github.com/openjdk/valhalla/pull/1117#discussion_r1640103814

tl;dr for this stage of the work, I didn't want to disable the existing `sum` intrinsic, but I think an `add` method works better in a larger context where methods for subtraction, multiplication, and division are being added.

**Comment by jddarcy** [comment]
> > > > > Hi @jddarcy , Apart from few minor comments, patch looks good to me, there is a build error due to malformed javadoc comment.
> > > > > Kindly fix and integrate.
> > > > 
> > > > 
> > > > Thanks; let me take a pass at writing at least some basic regression tests before pushing.
> > > > @jatin-bhateja , do you know if promoting the three operands of a Float16 fma to double, doing the operation in double, and rounding to Float16 is sufficient to correctly implement a Float16 fma? I haven't worked through all the cases yet and I'm not certain they're cannot be double-rounding issues. (If double rounding turns out to be a problem, I was thinking it would be possible to see if (a*b + c) was exact in double, and if, not add in a sticky bit to make sure the rounding occurs properly, but I haven't developed the details yet.)
> > > 
> > > 
> > > Hi @jddarcy , As per specification of [Math.fma(float, float, float)](https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/lang/Math.java#L2494) internal computation of constituent operation (mul and add) should be done at infinite precision and only final result should be rounded, we are now upcasting float16 to double but it will not prevent rounding happening for mul and add.
> > 
> > 
> > Right; the fma needs to operate as-if it used infinite precision internally. This could be implemented (slowly) using JDK classes by implementing a BigDecimal -> Float16 conversion after doing a*b+c in BigDeicmal. I've been considering adding BigDecimal -> Float16 conversion anyway for completeness in the platform.
> > My understanding of how fma is implemented in hardware is that for a format with P bits of precision, there is a ~2P wide internal register to hold the exact product as an intermediate result. Then the value being adding in can be aligned at the right exponent location and the final rounding back to P bits of precision can occur, with logic of any sticky bit for rounding to nearest even, etc.
> > There are many cases where double (P = 53) will exactly hold the product and sum of three Float16 (P = 11) operands. However, the product can be so large or so small that rounding occurs when the third operand is added in.
> > I haven't worked through if the potential round-offs are all benign with the final rounding to Float16 or if some corrective action would need to be taken to get the effect of a sticky bit. For example, if a_b is so large that the highest exponent position set is more than 53 positions away from the lowest exponent position set on c, if the final result is going to overflow anyway, the round-off in computing a_b+c in double doesn't matter. It might be problematic if a*b is much smaller than c, but that is another case I haven't fully thought through yet.
> > I'll give an updated on my analysis/research on this fma issue by next week.
> 
> An update: promoting the three float16 fma operands to double and then computing (a*b) + c in double and rounding once at the end to Float16 works for (at least) most possible operands to a Float16 fma. It may work for all operands; I'm still working through those details.
> 
> To summarize my findings. if the product-sum held in double is exact, the single rounding to Float16 will compute the correct result.
> 
> Each Float16 operand is exactly convertible to double and the a*b product is exact since double has more than twice the precision of Float16. That leaves analyzing whether or not the product-sum is exact.
> 
> For Float16 values, the possible exponent bit positions that could be set range from 2^15 (from MAX_VALUE) to 2^(-24) (from MIN_VALUE). This entire exponent range of 15 - (-24) + 1 = 40 is less than the _precision_ of double, 53 bits.
> 
> When multiplying two Float16 values together, the exponent bit positions set in the exact product range over 2^31 (MAX_VALUE squared) to 2^(-48) (MIN_VALUE) squared.
> 
> If the exact product is larger than about 2^16, the final rounded result must overflow Float16. If the exact product has a highest exponent bit of 2^16, the product-sum will be exact since a double could hold a value as small as 2^(16 - 52) = 2^(-36), which is smaller than Float16.MIN_VALUE. Therefore, when the product has an exponent of 2^16 or larger, the end result is correct after conversion to Float16.
> 
> A similar argument holds when the product has an exponent in the normal range of Float16; any additional sum will be exact given the precision of double.
> 
> However, if the product is sufficiently small, it is possible the double will not be able to hold the exact value since the exponent range of 2^(-48) to 2^(15) is larger than double precision. However, for many operands, even if the product has underflowed, adding in many operands may be exact in practice. This exactness can be tested for using the 2sum algorithm (https://en.wikipedia.org/wiki/2Sum) for a zero trailing "t" component.
> 
> Fortunately in terms of computing the final answer and rounding complications, if c is the subnormal range of Float16 (exponent range 2^(-14) to 2(-24)), the product-sum will be exactly representable in double.
> 
> That leaves possibly non-exact product-sum with a combination of product in the subnormal range of Float16 and the c term to be added in being not-small. However, if this product-sum is non-exact, the smaller term from the product, with at most 22 exponent bit positions set, and the the 11 bits from c being summed in, must be separated by at least 53 - (22 + 11) = 20 bit positions otherwise the product-sum would fit in a double. I believe this implies at least one of the double-rounding scenarios cannot occur, in particular a half-way result in the smaller precision, Float16 in this case, rounding differently because sticky bit information from the higher precision was rounded away.
> 
> I'll keep working through the non-exact cases. Assuming a `Float16 valueOf(BigDecimal)` method is added, an "obviously right" approach would be to use a*b + c in double for all exact cases and failover to computing the exact result in BigDecimal and then rounding to that to Float16 for all the inexact cases.

PS I issued an integrate command to checkpoint the current state of the work and allow collaboration on other aspects of it, like adding intrinsics for more methods. I filed https://bugs.openjdk.org/browse/JDK-8334432 to track refinements to to Float16.fma. If nothing else, even if the current implementation is correct, the analysis of that should be included as a code comment.

**Comment by jatin-bhateja** [comment]
Thanks @jddarcy , for detailed explanations. 
Sponsoring the patch.

**Comment by briangoetz** [comment]
> tl;dr for this stage of the work, I didn't want to disable the existing `sum` intrinsic, but I think an `add` method works better in a larger context where methods for subtraction, multiplication, and division are being added.

Good, that's what I was looking for.  One of the things we hope to gain consensus on in this experiment is conventions for naming and organizing operations on numerics.  There may be some inevitable duplication (because Number asks for a particular method name) but we need to establish the right conventions for going forward.
